server/base/src/main/java/org/apache/accumulo/server/metrics/MetricsInfoImpl.java

```diff
@@ -0,0 +1,340 @@
+  }
+
+  private void printMetricsConfig() {
+    final boolean micrometerEnabled =
+        context.getConfiguration().getBoolean(Property.GENERAL_MICROMETER_ENABLED);
+    final boolean jvmMetricsEnabled =
+        context.getConfiguration().getBoolean(Property.GENERAL_MICROMETER_JVM_METRICS_ENABLED);
+    LOG.info("micrometer metrics enabled: {}", micrometerEnabled);
+    if (jvmMetricsEnabled) {
+      if (micrometerEnabled) {
+        LOG.info("detailed jvm metrics enabled: {}", jvmMetricsEnabled);
+      } else {
+        LOG.info("requested jvm metrics, but micrometer metrics are disabled.");
+      }
+    }
+    if (micrometerEnabled) {
+      LOG.info("metrics registry factories: {}",
+          context.getConfiguration().get(Property.GENERAL_MICROMETER_FACTORY));
+    }
+  }
+
+  @Override
+  public boolean metricsEnabled() {
+    return metricsEnabled;
+  }
+
+  /**
+   * Common tags for all services.
+   */
+  @Override
+  public void addServiceTags(final String applicationName, final HostAndPort hostAndPort) {
+    List<Tag> tags = new ArrayList<>();
+
+    if (applicationName != null && !applicationName.isEmpty()) {
+      tags.add(MetricsInfo.processTag(applicationName));
+    }
+    if (hostAndPort != null) {
+      tags.addAll(MetricsInfo.addressTags(hostAndPort));
+    }
+    addCommonTags(tags);
+  }
+
+  @Override
+  public void addCommonTags(List<Tag> updates) {
+    lock.lock();
+    try {
+      if (composite != null) {
+        LOG.warn(
+            "Common tags after registry has been initialized may be ignored. Current common tags: {}",
+            commonTags);
+        return;
+      }
+      updates.forEach(t -> commonTags.put(t.getKey(), t));
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public Collection<Tag> getCommonTags() {
+    lock.lock();
+    try {
+      return Collections.unmodifiableCollection(commonTags.values());
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public void addRegistry(MeterRegistry registry) {
+    lock.lock();
+    try {
+      if (composite != null) {
+        composite.add(registry);
+      } else {
+        // defer until composite is initialized
+        pendingRegistries.add(registry);
+      }
+
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public void addMetricsProducers(MetricsProducer... producer) {
+    lock.lock();
+    try {
+      if (composite == null) {
+        producers.addAll(Arrays.asList(producer));
+      } else {
+        Arrays.stream(producer).forEach(p -> p.registerMetrics(composite));
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public MeterRegistry getRegistry() {
+    lock.lock();
+    try {
+      if (composite == null) {
+        throw new IllegalStateException("metrics have not been initialized, call init() first");
+      }
+    } finally {
+      lock.unlock();
+    }
+    return composite;
+  }
+
+  @Override
+  public void init() {
```

Comment:
It appears there is a common pattern being used by the server processes:

```
addServiceTags();
addMetricsProducers();
init()
```

I'm wondering if this could be collapsed to one method that initializes the MetricsInfo object and makes it ready for use.

Reply:
The pattern is common, but in `TimedProcessor` there is an example where a producer is added, but init() is not directly called.  With `TServerUtils` it was necessary to allow producers to be added, but init() deferred until service start.

I kept the methods separate because it is not required that they are always called back-to-back.  There could be an `confgAndInit()` method that would collapse the pattern, but addProducers() would still be necessary and seemed it could be more confusing to understand when to use one or the other, and the discrete methods made it more explicit,

Reply:
So you have a multi-step build process where some of the steps are optional. This seems like a good place for a Builder then. It seems that once `init` is called, then the configuration is fixed and ready for use, like a `build()` method that creates an un-modifiable object. I think the implementation here would be that ServerContext would have a reference to a MetricsInfoBuilder, it would be passed around  as needed to configure all the things, then the server process would call build() on it which would be much like the `init` method. Just a thought.

Reply:
That is a good observation and essentially correct.  I'd need to think about a builder and maybe do it as a follow-on.   I didn't consider adding a builder to server context and passing that around.

The key thing is that common tags must be bound to the registry before creating / binding meters.  

```
From the micrometer tag documentation:

Common tags generally have to be added to the registry before any (possibly autoconfigured) meter binders. Depending on your environment, there are different ways to achieve this.
```

That's what prompted this whole set of changes - to gather the common tags (some of which are run-time dependent (i.e. port number)) and then apply them to the registry and then perform the registration of the producers - sealing the common tags with the init() call.